src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -167,43 +157,149 @@ internal int EndRead(IAsyncResult asyncResult)
+                if (t.IsCompleted)
+                {
+                    _nestedWrite = 0;
+                    return t;
+                }
+                else
+                {
+                    return ExitWriteAsync(t);
+                }
+            }
+            else
+            {
+                return ExitWriteAsync(WriteAsyncChunked(writeAdapter, buffer, offset, count));
             }
 
-            // No "artificial" timeouts implemented so far, InnerStream controls timeout.
-            lazyResult.InternalWaitForCompletion();
-
-            if (lazyResult.Result is Exception e)
+            async Task ExitWriteAsync(Task task)
             {
-                if (e is IOException)
+                try
                 {
-                    ExceptionDispatchInfo.Throw(e);
+                    await task.ConfigureAwait(false);
                 }
+                catch (Exception e)
+                {
+                    _sslState.FinishWrite();
+
+                    if (e is IOException)
+                    {
+                        throw;
+                    }
 
-                throw new IOException(SR.net_io_write, e);
+                    throw new IOException(SR.net_io_write, e);
+                }
+                finally
+                {
+                    _nestedWrite = 0;
+                }
             }
         }
 
+        private Task WriteSingleChunk<T>(T writeAdapter, byte[] buffer, int offset, int count)
+            where T : struct, ISslWriteAdapter
+        {
+            // Request a write IO slot.
+            Task ioSlot = writeAdapter.LockAsync();
+            if (!ioSlot.IsCompletedSuccessfully)
+            {
+                // Operation is async and has been queued, return.
+                return WaitForWriteIOSlot(writeAdapter, ioSlot, buffer, offset, count);
+            }
+
+            byte[] rentedBuffer = ArrayPool<byte>.Shared.Rent(count + FrameOverhead);
+            byte[] outBuffer = rentedBuffer;
+
+            SecurityStatusPal status = _sslState.EncryptData(buffer, offset, count, ref outBuffer, out int encryptedBytes);
+
+            if (status.ErrorCode != SecurityStatusPalErrorCode.OK)
+            {
+                // Re-handshake status is not supported.
+                ArrayPool<byte>.Shared.Return(rentedBuffer);
+                ProtocolToken message = new ProtocolToken(null, status);
+                throw new IOException(SR.net_io_encrypt, message.GetException());
+            }
+
+            Task t = writeAdapter.WriteAsync(outBuffer, 0, encryptedBytes);
+            if (t.IsCompletedSuccessfully)
+            {
+                ArrayPool<byte>.Shared.Return(rentedBuffer);
+                _sslState.FinishWrite();
+                return t;
+            }
+            else
+            {
+                return CompleteAsync(t, rentedBuffer);
+            }
+
+            async Task WaitForWriteIOSlot<T2>(T2 wAdapter, Task lockTask, byte[] buff, int off, int c)
```

Comment:
Why does this need its own generic type parameter... doesn't it inherit the one from the containing method?

Reply:
i didn't realise that worked, changed.